File: java/Day02.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.io.File;
import java.io.FileNotFoundException;

public class Day02 {

    // Reads the file and generates a list of reports
    public static List<List<Integer>> getReports(String file) {
        List<List<Integer>> reports = new ArrayList<>();
        try (Scanner sc = new Scanner(new File(file))) {
            while (sc.hasNextLine()) {
                String line = sc.nextLine();
                String nums[] = line.split(" ");
                List<Integer> reportNums = new ArrayList<>();
                for (String num : nums) {
                    reportNums.add(Integer.parseInt(num));
                }
                reports.add(reportNums);
            }
        } catch (FileNotFoundException e) {
            System.out.println("File not found");
        }
        return reports;
    }

    // Part 1
    public static int getSafe(String file) {
        List<List<Integer>> reports = getReports(file);
        int safeCount = 0;

        for (List<Integer> report : reports) {
            boolean isSafe = true;
            boolean increasing = report.get(0) < report.get(1);

            for (int i = 1; i < report.size(); i++) {
                int diff = Math.abs(report.get(i - 1) - report.get(i));
                if (diff > 3 || diff < 1) {
                    isSafe = false;
                    break;
                }
                if (increasing && report.get(i) < report.get(i - 1)) {
                    isSafe = false;
                    break;
                } else if (!increasing && report.get(i) > report.get(i - 1)) {
                    isSafe = false;
                    break;
                }
            }
            if (isSafe) safeCount++;
        }

        System.out.println("Part 1: " + safeCount);
        return safeCount;
    }

    // Part 2
    public static int getSafePart2(String file) {
        List<List<Integer>> reports = getReports(file);
        int safeCount = 0;

        for (List<Integer> report : reports) {
            if (isReportSafe(report)) {
                safeCount++;
            } else {
                // Try removing each level and check if it becomes safe
                for (int i = 0; i < report.size(); i++) {
                    List<Integer> modifiedReport = new ArrayList<>(report);
                    modifiedReport.remove(i);
                    if (isReportSafe(modifiedReport)) {
                        safeCount++;
                        break;
                    }
                }
            }
        }

        System.out.println("Part 2: " + safeCount);
        return safeCount;
    }

    // Helper method to check if a report is safe
    private static boolean isReportSafe(List<Integer> report) {
        if (report.size() < 2) return true; 

        boolean increasing = report.get(0) < report.get(1);

        for (int i = 1; i < report.size(); i++) {
            int diff = Math.abs(report.get(i - 1) - report.get(i));
            if (diff > 3 || diff < 1) return false;
            if (increasing && report.get(i) < report.get(i - 1)) return false;
            if (!increasing && report.get(i) > report.get(i - 1)) return false;
        }
        return true;
    }

    public static void main(String[] args) {
        System.out.println(getSafe("C:\\Users\\abby-\\projects\\advent2024\\java\\input.txt"));
        System.out.println(getSafePart2("C:\\Users\\abby-\\projects\\advent2024\\java\\input.txt"));
    }
}

File: js/day02.js
const fs = require('fs');

function getReports(filename) {
    const lines = fs.readFileSync(filename).toString().split('\n');
    const reports = [];
    lines.forEach((line) => {
        const report = line.split(" ").map(num => parseInt(num));
        reports.push(report);
    });
    return reports;
}

function isSafe(report) {
    // if (report.length == 1) return true;
    const increasing = (report[0] - report[1]) < 0 ? true : false;
    for (let i = 1; i < report.length; i++) {
        const diff = Math.abs(report[i-1] - report[i]);
        if (diff > 3 || diff < 1) return false;
        if (increasing && report[i-1] - report[i] > 0) {
            return false;
        } else if (!increasing && report[i-1] - report[i] < 0) {
            return false;
        }
    }
    return true;
}

function getPartOne(filename) {
    const reports = getReports(filename);
    // console.log(reports);
    let safe = 0;
    // loop through the reports and count which ones are safe
    for (let i = 0; i < reports.length; i++) {
        // console.log(reports[i]);
        if (isSafe(reports[i])) {
            safe++;
        }
    }
    console.log("part 1: " + safe);
    return safe;
}

function getPartTwo(filename) {
    const reports = getReports(filename);
    let safe = 0;
    for (let i = 0; i < reports.length; i++) {
        if (isSafe(reports[i])) {
            safe++;
        } else {
            for (let j = 0; j < reports[i].length; j++) {
                const modReport = [...reports[i].slice(0, j), ...reports[i].slice(j + 1)];
                if (isSafe(modReport)) {
                    safe++;
                    break;
                }
            }
        }
    }
    console.log("part 2: " + safe);
}

getPartOne("C:\\Users\\abby-\\projects\\advent2024\\java\\input.txt");
getPartTwo("C:\\Users\\abby-\\projects\\advent2024\\java\\input.txt");